### PR TITLE
Fix a bug where product media upload via API was not possible anymore.

### PR DESCRIPTION
**Pull request by digitalpianism**
### Description (*)

The following commit (https://github.com/edannenberg/magento-lts/commit/cd0c30fe8bbd6d966e267081bca8b223f62e0ff2) has introduced a few bugs:
- https://github.com/OpenMage/magento-lts/issues/1178
- https://github.com/OpenMage/magento-lts/issues/1125

The purpose of this PR is to fix those bugs by trying to keep the original changes made in the commit I mentionned above.

### Fixed Issues (if relevant)

1. Fixes OpenMage/magento-lts#1178
2. Fixes OpenMage/magento-lts#1125
3. Fixes the EVIL issue OpenMage/magento-lts#666

### Manual testing scenarios (*)

1. Run the following php file from the root of your Magento website.
```
<?php
//some settings
error_reporting(E_ALL | E_STRICT);
define('MAGENTO_ROOT', getcwd());
$mageFilename = MAGENTO_ROOT . '/app/Mage.php';
require_once $mageFilename;
Mage::setIsDeveloperMode(true);
ini_set('display_errors', 1);
umask(0);
//instantiate the app model
Mage::app();
$tmpDirectory = Mage::getBaseDir('var') . DS . 'api' . DS . "test";
$ioAdapter = new Varien_Io_File();
try {
    // TEST 1: simple text file
    $fileContent = "test";
    $fileName  = 'test1.txt';
    // Create temporary directory for api
    $ioAdapter->checkAndCreateFolder($tmpDirectory);
    $ioAdapter->open(array('path' => $tmpDirectory));
    $ioAdapter->write($fileName, $fileContent, 0666);
    unset($fileContent);
} catch (Exception $e) {
    echo $e->getMessage() . "\n";
}

$ioAdapter = new Varien_Io_File();
try {
    // TEST 2: image file from URL
    $fileContent = file_get_contents("http://www.google.com/images/branding/googlelogo/2x/googlelogo_color_92x30dp.png");
    $fileName  = 'test2.png';
    // Create temporary directory for api
    $ioAdapter->checkAndCreateFolder($tmpDirectory);
    $ioAdapter->open(array('path' => $tmpDirectory));
    // Write image file
    $ioAdapter->write($fileName, $fileContent, 0666);
    unset($fileContent);
} catch (Exception $e) {
    echo $e->getMessage() . "\n";
}

$ioAdapter = new Varien_Io_File();
try {
    // TEST 3: image file from path
    $fileContent = $tmpDirectory . DS . "test2.png";
    $fileName  = 'test3.png';
    // Create temporary directory for api
    $ioAdapter->checkAndCreateFolder($tmpDirectory);
    $ioAdapter->open(array('path' => $tmpDirectory));
    // Write image file
    $ioAdapter->write($fileName, $fileContent, 0666);
    unset($fileContent);
} catch (Exception $e) {
    echo $e->getMessage() . "\n";
}

$ioAdapter = new Varien_Io_File();
try {
    // TEST 4: image file from URL with addition of base64 encode/decode
    // To reproduce the behavior of Mage_Catalog_Model_Product_Attribute_Media_Api::create()
    $imageFile = "http://www.google.com/images/branding/googlelogo/2x/googlelogo_color_92x30dp.png";
    $imData = file_get_contents($imageFile);
    $fileContentSent = base64_encode($imData);
    $fileContent = @base64_decode($fileContentSent, true);
    $fileName  = 'test4.png';
    // Create temporary directory for api
    $ioAdapter->checkAndCreateFolder($tmpDirectory);
    $ioAdapter->open(array('path' => $tmpDirectory));
    // Write image file
    $ioAdapter->write($fileName, $fileContent, 0666);
    unset($fileContent);
} catch (Exception $e) {
    echo $e->getMessage() . "\n";
}

$ioAdapter = new Varien_Io_File();
try {
    // TEST 5: image file from path with addition of base64 encode/decode
    // To reproduce the behavior of Mage_Catalog_Model_Product_Attribute_Media_Api::create()
    $imageFile = $tmpDirectory . DS . "test3.png";
    $imData = file_get_contents($imageFile);
    $fileContentSent = base64_encode($imData);
    $fileContent = @base64_decode($fileContentSent, true);
    $fileName  = 'test5.png';
    // Create temporary directory for api
    $ioAdapter->checkAndCreateFolder($tmpDirectory);
    $ioAdapter->open(array('path' => $tmpDirectory));
    // Write image file
    $ioAdapter->write($fileName, $fileContent, 0666);
    unset($fileContent);
} catch (Exception $e) {
    echo $e->getMessage() . "\n";
}

```
2. You will get only **3 files out of 5** under `var/api/test` folder. **Both images will be empty images**
3. Apply the PR 
4. Run the file again.
5. You will get **all the files** from the 5 different tests. **All images will be correct.**

### Questions or comments

The original purpose of the buggy commit was a security fix: prevent null byte inside filename from being processed when writing a file. The **problem introduced** was that the source ([as explained in the PHPDoc)](https://github.com/OpenMage/magento-lts/blob/1.9.4.x/lib/Varien/Io/File.php#L452) can be **either a string, a file or a stream**. Thus, what I've done is to **check if the string provided is a file, and in that case only, check for null byte**. 

Indeed, **when you use file_get_content as a source** for the Varien_Io_File::write method, **the original null byte check will return true and thus the write will fail silently.**

Note that I have also re added the @ operator to the is_readable call as it throw a warning in case you provide content directly instead of providing a file path.

### Contribution checklist (*)
 - [x] Pull request has a meaningful description of its purpose
 - [x] All commits are accompanied by meaningful commit messages
 - [ ] All automated tests passed successfully (all builds are green)

### CODE REVIEW

**Comment by colinmollenhour** [comment]
Looks good but please revert the unrelated changes to change log and Mage.php

**Comment by digitalpianism** [comment]
@colinmollenhour done sorry that was a mistake I didn't want to changes those two files ^^

**Review by colinmollenhour — DISMISSED** [review]

File: lib/Varien/Io/File.php [hunk 1]
```diff
@@ -486,8 +486,17 @@ public function write($filename, $src, $mode=null)
...
+        // In case of a string
+        if (is_string($src)) {
+            // If its a file we check for null byte
+            // If it's not a valid path, file_exists() will return a falsey value, and the @ will keep it from complaining about the bad string.
+            if (@file_exists($src)
+                && strpos($src, chr(0)) !== false) {
+                return false;
+            } else {
+                return true;
+            }
```
Comment (sreichel) [inline lib/Varien/Io/File.php:498]:
```suggestion
            return @file_exists($src) && strpos($src, chr(0)) !== false ? false : true;
```

**Review by sreichel — APPROVED** [review]

**Comment by maximehuran** [comment]
We applied this fix on a store which using API to create images and it works

**Comment by rjocoleman** [comment]
I encountered this bug today too! Thanks for the fix @digitalpianism

Along with the fix this PR brings a few things with it, they're all cancelled out but it will introduce 9 irrelevant commits that are then undone by another commit.

The fix itself is only this: https://gist.github.com/rjocoleman/4422ca5cd56fe5ca84a316170f7196c0

@digitalpianism to fix up the commit history and not introduce the extra cruft could you please rebase? 

(One way to do this by rebasing the branch interactively against the base branch and dropping all the irrelevant commits, then "squashing" the relevant ones together https://asciinema.org/a/JzWvAeKaFcY7h51mvcFXD1X8h - if you do this you'd need to then force push the `fix-1178` branch to overwrite the old history)

Aside from keeping the commit history clean a major advantage is being about to use things like [composer patches](https://github.com/cweagans/composer-patches) to apply the fix from this PR before it's merged (as it stands the patch is really large and might not cleanly apply because of the irrelevant commits)

thanks!

**Comment by wytzejan** [comment]
When can this one be merged? @digitalpianism 
I have also applied this fix on a store which using API to create images and it works. The only thing is that everytime I update Magento I manually have to make this code change in the core.

**Comment by rvelhote** [comment]
We were doing some data transfer using the API and were affected by this issue. This patch to `lib/Varien/Io/File.php` fixed the issue for us. No side-effects were detected.

**Review by colinmollenhour — APPROVED** [review]

**Pull request merged**